src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/DefaultShapeTableManager.cs

```diff
@@ -45,20 +46,28 @@ public class DefaultShapeTableManager : IShapeTableManager
 
         public Task<ShapeTable> GetShapeTableAsync(string themeId)
         {
-            // This method is intentionally kept non-async since most calls
+            // This method is intentionally not awaited since most calls
             // are from cache.
 
             if (_shapeTableCache.TryGetValue(themeId ?? DefaultThemeIdKey, out var shapeTable))
             {
                 return Task.FromResult(shapeTable);
             }
 
-            return BuildShapeTableAsync(themeId);
+            lock (_shapeTableCache)
```

Comment:
We should only have to lock if the collection is empty so we can populate it. Otherwise this lock will cause performance degrade.

Reply:
It is blocking anything from creating a new `ShapeTable`, true, this is done once per theme per tenant. In 99.9999% of the calls it's not even reaching this state. Then the lock is only blocking when there are two threads creating the tables, and the second one will just wait as long as it would have taken it to create the table. All good IMO.

Reply:
Isn't this code is called on every request? But since there is a cache, so it read from the cache on every request. So locking the cache here will lock/unlock on every request not just once per theme/per tenant. Locking once per theme/per tenant is good, but locking the cache per request is a problem IMO.

Reply:
I just looked at the updated code and it looks good. It only lock when no cache is found.